Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,628 @@
+                // refactoring or eliminating that code easier once the rust sync manager
+                // rollout is complete.
+                Scratchpad.clearFromPrefs(self.prefsForSync.branch("scratchpad"))
+                self.prefsForSync.clearAll()
+            }
+            return succeed()
+        }
+        syncQueue.async { [unowned self] in
+            self.syncManagerAPI.disconnect()
+        }
+        return clearPrefs()
+    }
+
+    private func getEngineEnablementChangesForAccount() -> [String: Bool] {
+        var engineEnablements: [String: Bool] = [:]
+        // We just created the account, the user went through the Choose What to Sync
+        // screen on FxA.
+        if let declined = UserDefaults.standard.stringArray(
+            forKey: "fxa.cwts.declinedSyncEngines") {
+            declined.forEach { engineEnablements[$0] = false }
+            UserDefaults.standard.removeObject(forKey: "fxa.cwts.declinedSyncEngines")
+        } else {
+            // Bundle in authState the engines the user activated/disabled since the
+            // last sync.
+            RustTogglableEngines.forEach { engine in
+                let stateChangedPref = "engine.\(engine).enabledStateChanged"
+                if self.prefsForSync.boolForKey(stateChangedPref) != nil,
+                   let enabled = self.prefsForSync.boolForKey("engine.\(engine).enabled") {
+                    engineEnablements[engine] = enabled
+                    self.prefsForSync.setObject(nil, forKey: stateChangedPref)
+                }
+            }
+        }
+
+        if !engineEnablements.isEmpty {
+            logger.log("""
+                engines to enable:
+                \(engineEnablements.compactMap { $0.value ? $0.key : nil })
+                """,
+               level: .info,
+               category: .sync)
+            logger.log("""
+                engines to disable:
+                \(engineEnablements.compactMap { !$0.value ? $0.key : nil })
+                """,
+               level: .info,
+               category: .sync)
+        }
+        return engineEnablements
+    }
+
+    public class ScopedKeyError: MaybeErrorType {
+        public let description = "No key data found for scope."
+    }
+
+    public class EncryptionKeyError: MaybeErrorType {
+        public let description = "Failed to get stored key."
+    }
+
+    public class DeviceIdError: MaybeErrorType {
+        public let description = "Failed to get deviceId."
+    }
+
+    public class NoTokenServerURLError: MaybeErrorType {
+        public let description = "Failed to get token server endpoint url."
+    }
+
+    public class EngineAndKeyRetrievalError: MaybeErrorType {
+        public let description = "Failed to get sync engine and key data."
+    }
+
+    fileprivate func getEnginesAndKeys(
+        engines: [String]
+    ) -> Deferred<Maybe<([EngineIdentifier], [String: String])>> {
```

Comment:
Looking at the code, I guess it would be too much work to remove Deferred usage? 🥲

Reply:
Yeah, because of the `SyncManager` protocol removing Deferred now would require a larger refactor. However, once the Nimbus sync manager integration rollout is over we can update the Deferred usage and refactor the callers.